Sc 1371 network v1 schema definition

blackboxsw opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

```
    schema: add network v1 schema definition and validation functions
    
    Part 1 of 3: Define network-config-v1 schema which is inert and not
    yet tied to either initial boot validation or the command:
    cloud-init schema.
    
    Part 2 will plumb this network schema validation to log warnings
    Part 3 will extend the command cloud-init schema to process and raise
    network validation errors from the commandline.
    
    In this branch:
      - extract all network version: 1 doc examples from inline documentation into
        the doc/examples directory to allow for unittest coverage
      - Use jsonschema.exceptions.best_match improving anyOf, allOf errors
        ValidationErrors with specific details avoiding the unhelpful
        message: "is not valid under any of the given schemas"
      - Add static schema-cloud-config-v1.json to represent supported
        network config version 1
      - validate_cloudconfig_schema and get_schema now accept a new param
        config_type to select either network-config cloud-config schemas
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [ ] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/contributing.html)
 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

TheRealFalcon commented on cloudinit/config/schema.py:
```diff
@@ -249,9 +264,12 @@ def _anyOf(
 
     all_errors = []
     all_deprecations = []
+    errors_for_declared_type = []
```
Unused variable

TheRealFalcon commented on cloudinit/config/schema.py:
```diff
@@ -263,8 +281,16 @@ def _anyOf(
             break
+        if "type" in instance and "anyOf_type" in subschema.get("$ref", ""):
+            if f"anyOf_type_{instance['type']}" in subschema["$ref"]:
+                # A matching anyOf_type_XXX $ref indicates this is likely the
```
Why is this likely the best match? Did something about network validation make this change, or is this just an enchancement? I'm having a hard time grokking this function in general, but that's just my lack of understand of json schema.

blackboxsw replied:
Really this is a shortfall of our abundant use of anyOf conditions as it is hard to report 'the right' error in many cases. It's also a shortfall of network version 1 definitions which allowed 6 different potential objects in the same list differentiated only by a nested `type:` key within the dict. So, our JSON schema is now trying to cleanly represent an all too flexible schema w/ `anyOf` clauses which reduce our JSONschema's effectiveness in determining specific errors.

JSON schema's `best_match` algorithm which only says the [deepest match into the match path `anyOf` structures is likely the best match[(https://python-jsonschema.readthedocs.io/en/latest/errors/#best-match-and-relevance). But, since our `anyOf` subschemas for `type: (physical|bridge|bond|vlan)` all that happen to fail at the same depth, JSON schema's `best_match` algorithm chooses the wrong (first) item to report "most likely" errors from. 

This additional logic is an enhancement to the best match algorithm, because our instance["type"] is (bond|bridge|physical .. .etc) and our defined schema `$ref` name  happens to have encoded the specific type in the subschema name. If our current `instance['type']` == `bridge` and  `subschema[$ref] == anyOf_type_bridge`, we know errors in this subschema are the best matched errors for this object (as opposed to schema errors for `anyOf_type_vlan` which would likely just be `type is 'bridge' is not  equal to 'vlan'` among other vlan-specific required properties).

To give us more of a hint into the network types, I added a schema naming construct, which definitely needs more informative comments. Any schema reference name containing the prefix `anyOf_type_<type_name>` is a hint to our error matching logic that the subschema is intended to match `type: <type_name>`  objects.

TheRealFalcon replied:
Very helpful, thanks! Any chance most of that comment could make its way into the docstring? I have a feeling I'm gonna be staring at this code in 6 months wondering what I'm looking at.

blackboxsw replied:
done.

TheRealFalcon commented on cloudinit/config/schema.py:
```diff
@@ -658,6 +690,7 @@ def validate_cloudconfig_file(
     config_path: str,
     schema: dict,
     annotate: bool = False,
+    #   schema_type: Optional[str]: "user-data",
```
Commented code

blackboxsw replied:
for a subsequent PR. Dropping

TheRealFalcon commented on cloudinit/config/schema.py:
```diff
@@ -1170,34 +1203,24 @@ def get_schema_dir() -> str:
-    # If we change the location of versions.schema.json in github, we need
-    # to provide an updated PR to
-    # https://github.com/SchemaStore/schemastore.
+def get_schema(schema_type: Optional[str] = "cloud-config") -> dict:
```
Is `schema_type` actually Optional? We're using it as an index into `SCHEMA_FILES_BY_TYPE`, so if it is `None` we'll get a KeyError.

TheRealFalcon commented on cloudinit/config/schema.py:
```diff
@@ -1170,34 +1203,24 @@ def get_schema_dir() -> str:
+    Return empty schema when no specific schema file exists.
+    """
+    schema_file = os.path.join(
+        get_schema_dir(), SCHEMA_FILES_BY_TYPE[schema_type]["v1"]
```
What would happen if we ever had a `v2` version of the schema? Would we have to update this here as well? If so, I'm questioning the value of the "v1" keys on lines 54 and 55. If you think they are valuable, I wonder if it's also worth having a "latest" key.

blackboxsw replied:
Probably worth defining the `latest` key.  I think we'll need to differentiate v1 from v2 as we continue to support both. But, let's establish a latest alias that'll point to v1 until we add v2. At that point latest == v2.
We will have to change get_schema in a subsquent PR to add an optional preferred version because we do support network v1 and v2. But that work wasn't included in this PR.

TheRealFalcon reviewed (changes requested):
Really nice addition here. The doc and test updates are great! I left comments inline.

It miiiiight be worth moving the non-network related bits to a separate PR, just because they seem significant on their own, but only if you think it's worthwhile.

blackboxsw commented on cloudinit/config/schemas/schema-network-config-v1.json:
```diff
@@ -0,0 +1,562 @@
+            "params": {
+              "desciption": "The ``params`` key in a bond holds a dictionary of bonding parameters. This dictionary may be empty. For more details on what the various bonding parameters mean please read the Linux Kernel :file:`Bonding.txt`.",
+              "additionalProperties": false,
+              "properties":{
```
Our unittests have both [hyphenated](https://github.com/canonical/cloud-init/blob/main/tests/unittests/test_net.py#L170) and [underscore delimited](https://github.com/canonical/cloud-init/blob/main/tests/unittests/test_net.py#L2827) key names sprinkled throughout test_net.py.  We also have a number of previous ubuntu-sru test validation scripts verifying that both hyphenated and underscore-delimited keys are honored.
https://github.com/cloud-init/ubuntu-sru/blob/main/bugs/lp-1784699.txt#L39
bugs/lp-1690480/network_data.json.

This list is likely stale from a few years ago and I pulled the initial list form our advertised docs  [for network v1 ](https://cloudinit.readthedocs.io/en/latest/reference/network-config-format-v1.html#bond). I thought we'd start with what we published in docs and grow it as needed to reflect additional/newer key/config changes.

TheRealFalcon replied:
I'm a little wary of the day warnings become errors as any bonding param would have worked here, and it may turn into us needlessly restricting it and throwing errors for valid options. I think this is fine as-is, but I think it's worth thinking about what we'll need to change between now and then.

blackboxsw commented on cloudinit/config/schemas/schema-network-config-v1.json:
```diff
@@ -0,0 +1,562 @@
+                  "description": "Set the bridge's hello value."
+                },
+                "bridge_hw": {
+                  "type": "integer",
```
yep thanks.

blackboxsw commented on cloudinit/config/schemas/schema-network-config-v1.json:
```diff
@@ -0,0 +1,562 @@
+        "anyOf_type_route": {
+          "type": "object",
+          "additionalProperties": false,
+	  "anyOf": [
```
Thanks to @holmanb's `tools/check_json_format.sh` I've reformatted and added this to our `.github/workflows/check_format.yml` CI check as well.

blackboxsw commented on cloudinit/config/schema.py:
```diff
@@ -445,6 +474,8 @@ def validate_cloudconfig_schema(
     @param schema: jsonschema dict describing the supported schema definition
        for the cloud config module (config.cc_*). If None, validate against
        global schema.
+    @param schema_type: Optional string. One of: userdata, network_config
```
It probably should. Will change it.

blackboxsw commented on cloudinit/config/schemas/schema-network-config-v1.json:
```diff
@@ -0,0 +1,562 @@
+          "type": "object",
+          "additionalProperties": false,
+	  "anyOf": [
+	      {"required": ["network", "gateway"]},
```
Network or destination keys comes from https://github.com/canonical/cloud-init/blob/main/cloudinit/net/network_state.py#L1051-L1070 which will [raise ValueError if neither destination nor network is present](https://github.com/canonical/cloud-init/blob/main/cloudinit/net/network_state.py#L992-#L998)

blackboxsw replied:
Wihle it would be nice to make it required the python logic itself allows for no `type: route` definition to cut down on repetition when as the the parent key scope is typically `routes:`.
- [code comments suggest the optional nature of `type` key](url)
- OLD SRU unittests/integration/verification tests[[1](https://github.com/canonical/cloud-init/blob/main/tests/unittests/cmd/devel/test_net_convert.py#L204-L208)], [[2](https://github.com/canonical/cloud-init/blob/main/tests/unittests/test_net.py#L2253-L2255)], [[3](https://github.com/cloud-init/ubuntu-sru/blob/main/bugs/lp-openstack-ipv6-slaac-dhcpv6.txt#L798-L800)] also assert success without the `type: route` dict item.

blackboxsw commented on cloudinit/config/schemas/schema-network-config-v1.json:
```diff
@@ -0,0 +1,562 @@
+            }
+          }
+        },
+        "config_type_subnet": {
```
No, only because subnet is nested under many other keys and the `type:` key is optional in many cases and SRU verification runs.

blackboxsw commented on cloudinit/config/schemas/schema-network-config-v1.json:
```diff
@@ -0,0 +1,562 @@
+            }
+        }
+    },
+    "required": [
```
Hrm, right, Warn as deprecated? Or shall we just contine to permit network-config without top-level `network` key?

TheRealFalcon replied:
Yeah, this one is a little tougher since it is coming from the clouds. Users will see a deprecation warning but can't do anything to fix it. I'm leaning towards just permitting it even though I don't like it.

TheRealFalcon commented:
Is this ready for re-review?

TheRealFalcon reviewed (commented):
There's a test failure with the new schema:
`bond-lacp-rate: fast` isn't part of the schema.

I'm still wary of trying to specify every bond option rather than passing everything through. I know schema validation is just a warning now, but if we ever turn it into an error, it seems fairly likely we'll cause regressions as we probably haven't caught every valid option across every possible supported distro.

I'll defer to your judgement here though.

One other really odd thing I'm seeing. If I do
```
$ git fetch blackboxsw
$ git checkout blackboxsw/SC-1371-network-v1-schema-definition
$ git diff main -- tests/unittests/cmd/test_clean.py        
```
I get
```diff
diff --git a/tests/unittests/cmd/test_clean.py b/tests/unittests/cmd/test_clean.py
index 02dce7ab7..c5385b790 100644
--- a/tests/unittests/cmd/test_clean.py
+++ b/tests/unittests/cmd/test_clean.py
@@ -123,15 +123,12 @@ class TestClean:
         symlink = clean_paths.cloud_dir.join("mylink")
         sym_link(dir1.strpath, symlink.strpath)
 
-        with mock.patch.object(
-            cloudinit.settings, "CLEAN_RUNPARTS_DIR", clean_paths.clean_dir
-        ):
-            retcode = wrap_and_call(
-                "cloudinit.cmd.clean",
-                {"Init": {"side_effect": init_class}},
-                clean.remove_artifacts,
-                remove_logs=False,
-            )
+        retcode = wrap_and_call(
+            "cloudinit.cmd.clean",
+            {"Init": {"side_effect": init_class}},
+            clean.remove_artifacts,
+            remove_logs=False,
+        )
         assert 0 == retcode
         for path in (dir1, symlink):
             assert path.exists() is False, f"Unexpected {path} found"
@@ -149,15 +146,12 @@ class TestClean:
         for _dir in dirs:
             ensure_dir(_dir)
 
-        with mock.patch.object(
-            cloudinit.settings, "CLEAN_RUNPARTS_DIR", clean_paths.clean_dir
-        ):
-            retcode = wrap_and_call(
-                "cloudinit.cmd.clean",
-                {"Init": {"side_effect": init_class}},
-                clean.remove_artifacts,
-                remove_logs=False,
-            )
+        retcode = wrap_and_call(
+            "cloudinit.cmd.clean",
+            {"Init": {"side_effect": init_class}},
+            clean.remove_artifacts,
+            remove_logs=False,
+        )
         assert 0 == retcode
         for expected_dir in dirs[:2]:
             assert expected_dir.exists() is True, f"Missing {expected_dir}"
@@ -177,16 +171,13 @@ class TestClean:
         for _dir in dirs:
             ensure_dir(_dir)
 
-        with mock.patch.object(
-            cloudinit.settings, "CLEAN_RUNPARTS_DIR", clean_paths.clean_dir
-        ):
-            retcode = wrap_and_call(
-                "cloudinit.cmd.clean",
-                {"Init": {"side_effect": init_class}},
-                clean.remove_artifacts,
-                remove_logs=False,
-                remove_seed=True,
-            )
+        retcode = wrap_and_call(
+            "cloudinit.cmd.clean",
+            {"Init": {"side_effect": init_class}},
+            clean.remove_artifacts,
+            remove_logs=False,
+            remove_seed=True,
+        )
         assert 0 == retcode
         assert (
             clean_paths.cloud_dir.exists() is True

```

I'm seeing this even though `test_clean.py` doesn't show up as part of this PR. It's causing local test failures for me. Any idea what's going on there?

blackboxsw commented:
I believe I have resolved all comments. Thanks @TheRealFalcon

blackboxsw commented:
> There's a test failure with the new schema: `bond-lacp-rate: fast` isn't part of the schema.
> 

should have been `bond-lacp_rate` to align with https://www.kernel.org/doc/Documentation/networking/bonding.txt  which specified `lacp_rate`. There were conflicts in docs too. corrected schema and unittests here.

> I'm still wary of trying to specify every bond option rather than passing everything through. I know schema validation is just a warning now, but if we ever turn it into an error, it seems fairly likely we'll cause regressions as we probably haven't caught every valid option across every possible supported distro.
> 
> I'll defer to your judgement here though.

I want to keep the bond option docs because I'd llke to generate network docs from schema in a followup PR. I also think we can avoid "regressions" by treated network schema validation issues as Warnings instead of strict errors, even if we make cloud-config userdata strict and error hard, I don't want to do the same yet with network for the reasons/risk you state.


> I'm seeing this even though `test_clean.py` doesn't show up as part of this PR. It's causing local test failures for me. Any idea what's going on there?

Fixed through git rebase

TheRealFalcon reviewed (approved):
Good work!
